### PR TITLE
Feat/login with google

### PR DESCRIPTION
**Pull request by ngomy2601**

### CODE REVIEW

File: src/main/java/com/example/bookstoreproject/api/auth/AuthController.java [hunk 1]
```diff
@@ -36,4 +40,17 @@ public JwtTokenResponseDTO login(final @RequestBody LoginDTO loginDTO) throws Ex
             throw supplyUnauthorizedError().get();
         }
     }
+
+    @Operation(summary = "User login by google account")
+    @PostMapping("/google")
+    public JwtTokenResponseDTO loginGoogle(@RequestBody TokenRequestDTO tokenRequestDTO) {
```
Comment (voquanghoa) [inline src/main/java/com/example/bookstoreproject/api/auth/AuthController.java]:
```suggestion
    public JwtTokenResponseDTO loginGoogle(final @RequestBody TokenRequestDTO tokenRequestDTO) {
```

File: src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    private final RoleStore roleStore;
+
+    public UserDetails loginGoogle(final String decodedToken) {
+        final GoogleTokenPayload googleAccount = googleTokenVerifierService.googleIdTokenVerifier(decodedToken);
+        return userStore.findByUsername(googleAccount.getEmail())
+                .map(this::getJwtUserDetails)
+                .orElseGet(() -> createNewGoogleUser(googleAccount));
+    }
+
+    private JwtUserDetails getJwtUserDetails(User user) {
```
Comment (voquanghoa) [inline src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java]:
```suggestion
    private JwtUserDetails getJwtUserDetails(final User user) {
```

File: src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        final String roleName = roleStore.findRoleName(user.getRoleId());
+        return new JwtUserDetails(user.getId(), user.getUsername(), user.getPassword(), createAuthorities(roleName));
+    }
+
+    private JwtUserDetails createNewGoogleUser(GoogleTokenPayload googleAccount) {
```
Comment (voquanghoa) [inline src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java]:
```suggestion
    private JwtUserDetails createNewGoogleUser(final GoogleTokenPayload googleAccount) {
```

File: src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        final UUID roleId = roleStore.findIdByName("CONTRIBUTOR");
+        final User newUser = createUser(googleAccount, roleId);
+        return new JwtUserDetails(newUser.getId(), newUser.getUsername(), newUser.getPassword(), createAuthorities("CONTRIBUTOR"));
+    }
+
+    private User createUser(GoogleTokenPayload googleAccount, UUID roleId) {
```
Comment (voquanghoa) [inline src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java]:
```suggestion
    private User createUser(final GoogleTokenPayload googleAccount, final UUID roleId) {
```

**Review by voquanghoa — APPROVED** [review]

File: src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java (hunk 4 above)
Comment (ngomy2601) [reply src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java]: DONE

File: src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java (hunk 3 above)
Comment (ngomy2601) [reply src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java]: DONE

File: src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java (hunk 2 above)
Comment (ngomy2601) [reply src/main/java/com/example/bookstoreproject/domain/auths/GoogleLoginService.java]: DONE

File: src/main/java/com/example/bookstoreproject/api/auth/AuthController.java (hunk 1 above)
Comment (ngomy2601) [reply src/main/java/com/example/bookstoreproject/api/auth/AuthController.java]: DONE

**Pull request merged**